### ISSUE TITLE
lowerBin() and upperBin() parameter inconsistencies in a query

### ISSUE THREAD

**Issue by jakemulf** [deephaven/deephaven-core#1532]
**Description**

The following python query causes a `Cannot find method lowerBin(io.deephaven.db.tables.utils.DBDateTime, java.lang.String)` error

```
from deephaven.DBTimeUtils import expressionToNanos, convertDateTime
from deephaven.TableTools import newTable, dateTimeCol

bin = '00:02:00'
dateTime = convertDateTime("2020-01-01T00:35:01 NY")

source = newTable(
    dateTimeCol("DateTimes", dateTime)
)

result = source.update("DateTimesBinned = lowerBin(DateTimes, bin)")
``

However, the following query which one would expect to be equivalent, runs properly

```
from deephaven.DBTimeUtils import expressionToNanos, convertDateTime
from deephaven.TableTools import newTable, dateTimeCol

dateTime = convertDateTime("2020-01-01T00:35:01 NY")

source = newTable(
    dateTimeCol("DateTimes", dateTime)
)

result = source.update("DateTimesBinned = lowerBin(DateTimes, '00:02:00')")
```

**Expected results**

Both queries execute with no issue

**Actual results**

The query with the variable throws an error

**Additional details and attachments**

The same thing happens in groovy

**Comment by nbauernfeind** [comment]
The story here is grim. There are three related tasks 1) implicitly convert strings to DateTime values, 2) find the method to call, 3) apply query parameters to the formula.

The implicit conversion to DateTime is really only half-baked; it only works on string literals that exist within the formula. It does not support implicit conversion of variables. The code that decides which method is being called only knows the types of variables and not the values.

If we do allow implicit conversions of variables, then this these implicit conversions must occur while deciding whether or not a method is a candidate. We likely would select the method that uses the fewest of such implicit conversions.

It might be worth doing to give the user the ability to add implicit conversions to the QueryLibrary.

**Comment by rcaudy** [comment]
We should be able to get things right with a literal or a `QueryScope` variable. More or less everything else requires the user to tell us what the type is, or trigger conversion themselves. The problem here is with the bin interval, not the date time argument, right?

**Comment by nbauernfeind** [comment]
Yes the issue is with the bin interval. When it is inline, we attempt to replace it before the query parser runs. This replacement changes a string to a long. When it is passed in as a constant parameter, the query parsing only knows the types of parameters available and not their values. 

The hard part is giving the QueryParameter the ability to try conversions on values of the query parameters. Allowing the user to add their own implicit conversions is practically free after that.
